src/index.js

```diff
@@ -0,0 +1,202 @@
+/**
+  @typedef {(to: number | string, x?: { state: object, replace?: boolean }) => void} Navigate
+  @typedef {{
+    location: { pathname: string, search: string, hash: string, state?: object },
+    navigate: Navigate
+  }} Location
+  @typedef {{ basePath: string, baseParams: object, navigate: Navigate }} Base
+*/
+/**
+  @template {unknown} T
+  @typedef {Parameters<(routePath: string, createChildren: ((params: object) => T) | T) => void>} Route<T>
+*/
+
+import { createHistory } from './history'
+import { pick, callOrReturn } from './matching'
+
+/** @type {React.Context<Location | undefined>} */
+const locationContext = React.createContext(undefined)
+/** @type {React.Context<Base | undefined>} */
+const baseContext = React.createContext(undefined)
+
+const inBrowser = typeof window !== 'undefined'
+
+// TODO: eslint plugin for key warning of pairs
+/** @returns {{
+  routes<T>(...routes: Array<Route<T>>): JSX.Element,
+  route<T>(...route: Route<T>): JSX.Element
+}}
+*/
+export function useRouting({ initialLocation = undefined } = {}) {
+  return {
+    routes,
+    route(...route) { return routes(route) },
+  }
```

Comment:
Can we add something that allows us to match routes to non-react stuff?

```js
const { match } = useRouting()
const x = match(['a', 1], ['b', 2])
```
Could be a good idea to move this to another hook